File: ts-sdk/src/alerts/send.alerts.ts
import { AxiosInstance } from "axios";
import { Finding, isPrivateFindings } from "../findings";
import { GetFortaApiHeaders, GetFortaApiUrl, assertExists } from "../utils";
import { EntityType } from "../labels";

export type SendAlerts = (
  input: SendAlertsInput[] | SendAlertsInput
) => Promise<SendAlertsResponse[]>;

export function provideSendAlerts(
  axios: AxiosInstance,
  isProd: boolean,
  getFortaApiUrl: GetFortaApiUrl,
  getFortaApiHeaders: GetFortaApiHeaders
): SendAlerts {
  assertExists(axios, "axios");
  assertExists(getFortaApiUrl, "getFortaApiUrl");
  assertExists(getFortaApiHeaders, "getFortaApiHeaders");

  return async function sendAlerts(input: SendAlertsInput[] | SendAlertsInput) {
    if (!Array.isArray(input)) {
      input = [input];
    }

    const mutation = getMutationFromInput(input);
    // dont make the http call when running in development
    if (!isProd) {
      return [];
    }

    const response: RawGraphqlSendAlertsResponse = await axios.post(
      getFortaApiUrl(),
      mutation,
      getFortaApiHeaders()
    );

    if (response.data && response.data.errors)
      throw Error(JSON.stringify(response.data.errors));

    // TODO check for any partial errors and surface them (maybe mark the finding for retry?)
    return response.data.data.sendAlerts.alerts;
  };
}

export interface SendAlertsInput {
  botId: string;
  finding: Finding;
}

export interface RawGraphqlSendAlertsResponse {
  data: {
    data: {
      sendAlerts: {
        alerts: SendAlertsResponse[];
      };
    };
    errors: any;
  };
}

export interface SendAlertsResponse {
  alertHash?: string;
  error?: SendAlertError;
}

export interface SendAlertError {
  code: string;
  message: string;
}

const getMutationFromInput = (inputs: SendAlertsInput[]) => {
  return {
    query: `
      mutation SendAlerts(
        $alerts: [AlertRequestInput!]!
      ) {
        sendAlerts(alerts: $alerts) {
          alerts {
            alertHash
            error {
              code
              message
            }
          }
        }
      }
    `,
    variables: {
      alerts: inputs.map((input) => {
        const jsonFinding = JSON.parse(input.finding.toString());
        // convert enums to all caps to match graphql enums
        jsonFinding.type = jsonFinding.type.toUpperCase();
        jsonFinding.severity = jsonFinding.severity.toUpperCase();
        for (const label of jsonFinding.labels) {
          // TODO investigate why sometimes the entityType is an integer enum (instead of string)
          if (typeof label.entityType == "number") {
            label.entityType = EntityType[label.entityType];
          }
          label.entityType = label.entityType.toUpperCase();
        }
        // remove protocol field (not part of graphql schema)
        delete jsonFinding["protocol"];
        // remove any empty-value fields
        for (const key of Object.keys(jsonFinding)) {
          if (isEmptyValue(jsonFinding[key])) {
            delete jsonFinding[key];
          } else if (key === "labels") {
            // if there are labels, remove empty-value fields from them too
            for (const label of jsonFinding.labels) {
              for (const labelKey of Object.keys(label)) {
                if (isEmptyValue(label[labelKey])) {
                  delete label[labelKey];
                }
              }
            }
          }
        }
        // set private flag
        jsonFinding.private = isPrivateFindings();

        return {
          botId: input.botId,
          finding: jsonFinding,
        };
      }),
    },
  };
};

function isEmptyValue(val: any): boolean {
  if (val == null || val == undefined) return true;
  if (Array.isArray(val)) return val.length == 0;
  if (typeof val === "string") return val.length == 0;
  if (typeof val === "object") return Object.keys(val).length == 0;
  return false;
}

File: py-sdk/src/forta_bot_sdk/alerts/send_alerts.py
from typing import Callable, TypedDict
import json

from aiohttp import ClientSession
from ..utils import assert_exists, GetFortaApiUrl, GetFortaApiHeaders, GetAioHttpSession, Logger
from ..findings import Finding


class SendAlertsInput(TypedDict):
    bot_id: str
    finding: Finding


class SendAlertError(TypedDict):
    code: str
    message: str


class SendAlertsResponse(TypedDict):
    alert_hash: str
    error: SendAlertError


SendAlerts = Callable[[list[SendAlertsInput]
                       | SendAlertsInput], SendAlertsResponse]


def provide_send_alerts(
        get_aiohttp_session: GetAioHttpSession,
        is_prod: bool,
        get_forta_api_url: GetFortaApiUrl,
        get_forta_api_headers: GetFortaApiHeaders,
        logger: Logger) -> SendAlerts:
    assert_exists(get_aiohttp_session, 'get_aiohttp_session')
    assert_exists(get_forta_api_url, 'get_forta_api_url')
    assert_exists(get_forta_api_headers, 'get_forta_api_headers')
    assert_exists(logger, 'logger')

    async def send_alerts(input: list[SendAlertsInput] | SendAlertsInput) -> list[SendAlertsResponse]:
        if not type(input) == list:
            input = [input]

        mutation = get_mutation_from_input(input)
        # dont make the http call when running in development
        if not is_prod:
            return []

        session: ClientSession = await get_aiohttp_session()
        response = await session.post(
            get_forta_api_url(),
            json=mutation,
            headers=get_forta_api_headers())

        if response.status == 200:
            send_alerts_response = (await response.json()).get('data').get('sendAlerts').get('alerts')
            # TODO check for any errors and surface them (and mark the finding for retry?)
            return [{**item, 'alert_hash': item.get('alertHash')} for item in send_alerts_response]
        else:
            logger.debug(
                f'send_alerts received non-200 response code: {response.status}')
            raise Exception(await response.text())

    return send_alerts


def get_mutation_from_input(inputs: list[SendAlertsInput]) -> dict:
    mutation = """
  mutation SendAlerts($alerts: [AlertRequestInput!]!) {
      sendAlerts(alerts: $alerts) {
          alerts {
              alertHash
              error {
                  code
                  message
              }
          }
      }
  }
  """
    alerts = []
    # serialize the inputs list
    for input in inputs:
        # serialize finding to json
        json_finding = input["finding"].to_json()
        # convert enums to all caps to match graphql enums
        json_finding["type"] = json_finding["type"].upper()
        json_finding["severity"] = json_finding["severity"].upper()
        for label in json_finding.get("labels", []):
            label["entityType"] = label["entityType"].upper()
        # remove protocol field (not part of graphql schema)
        if 'protocol' in json_finding:
            del json_finding["protocol"]
        alerts.append({
            "botId": input["bot_id"],
            "finding": json_finding
        })

    return {
        'query': mutation,
        'variables': {
            'alerts': alerts
        }
    }
